allennlp/data/data_loaders/multi_process_data_loader.py

```diff
@@ -293,42 +336,43 @@ def _start_instance_workers(self, queue: mp.JoinableQueue, ctx) -> List[BaseProc
             workers.append(worker)
         return workers
 
-    def _join_workers(self, workers: List[BaseProcess]) -> None:
+    def _start_batch_workers(self, queue: mp.JoinableQueue, ctx) -> List[BaseProcess]:
+        workers: List[BaseProcess] = []
+        for worker_id in range(self.num_workers):
+            worker: BaseProcess = ctx.Process(
+                target=self._batch_worker, args=(worker_id, queue), daemon=True
+            )
+            worker.start()
+            workers.append(worker)
+        return workers
+
+    def _join_workers(self, workers: List[BaseProcess], queue) -> None:
+        # Each worker will be blocking on a call to `queue.join()`,
+        # calling `queue.task_done()` times the number of workers will
+        # call the `queue.join()` to return, and each worker should exit on its own.
+        for _ in range(len(workers)):
+            try:
+                queue.task_done()
+            except ValueError:
+                # This happens if a worker died early.
+                break
```

Comment:
Don't we have to do something about that? How are we propagating the error from the failed worker?

Reply:
This is executed in a `finally` cause. The error is still propagated.

Reply:
:+1: